### ISSUE TITLE
Create an instantiable, generic column

### ISSUE THREAD

**Issue by multimeric** [jtablesaw/tablesaw#924]
Currently it is not possible to create a column of an arbitrary type. `AbstractColumn<T>` is very close to being not abstract, and indeed the JavaDoc suggests that `emptyCopy()` is the only remaining abstract method for `AbstractColumn`. I'm not 100% sure what this method does, but might it be possible to have a default implementation for this so that we can directly instantiate generic columns?

**Comment by multimeric** [comment]
The motivation for this, by the way, is being able to store arbitrary data within the `Table`, that can be sliced out using the same indexing mechanisms as usual. For example we might want to store a denormalised list of aliases for each person in one of the columns. It isn't actually that important that we can *manipulate* such a column in the table, just that we can store and then retrieve it.

**Comment by benmccann** [comment]
I'm not quite understanding the use case. A list of aliases could be a `StringColumn`. You can treat it as just a `Column` after it's created if you want to handle it in a generic way

**Comment by lwhite1** [comment]
I'm afraid I don't understand either. Are you suggesting something like a Column<Object> or a column that can contain multiple elements in each cell (Column<List<String>> or arbitrary objects like Column<Person>?

**Comment by multimeric** [comment]
>  Are you suggesting something like a Column or a column that can contain multiple elements in each cell (Column<List> or arbitrary objects like Column?

I'm exactly suggesting "a column that can contain multiple elements in each cell". With my example for the list of aliases, I meant that each *cell* contains a list of strings, which is different to a `StringColumn` which contains only a single `String` for each cell. So, for this scenario I was hoping to be able to instantiate a `Column<ArrayList<String>>`.

For the same use-case I also wanted to store the path to a certain file pertaining to each row, so I would like a `Column<Path>`. In table format, this would look something like:

`StringColumn countryName` | `Column<ArrayList<String>> aliases` | `Path economicDataFile`
--- | --- | ---
`"Commonwealth of Australia"` | `{"Australia", "AU"}` | `"/path/to/file.txt"`
`"Republic of India"` | `{"India", "IN"}` | `"/path/to/file_2.txt"`

I hope you can see the benefits of such a flexible table, even if we don't explicitly provide methods for these generic columns. Although these could come later for a `ListColumn` etc.

**Comment by lwhite1** [comment]
Yeah, I can see the benefits. There's a closed issue for this idea that i opened years ago, and at one time it was possible for users to create new column types. 

It turns out that it is 
- difficult to write efficient code to do basic table operations with arbitrary column data types (e.g. you could use generics for everything, but might autobox all the primitive values everytime you executed a query), 
- difficult to modify the library without breaking the ability to create arbitrary column types, and 
- of limited value unless you're going to write code for reading, writing, displaying, sorting, etc. tables with columns like this. 

It's not impossible, but it would require a good engineer who was dedicated to making it happen. Do you know anyone like that?

**Comment by multimeric** [comment]
Okay, I take your point about the usefulness of generic columns being limited without explicit implementations of reading, writing, unboxing etc. I'm aware of this, but still think it's worthwhile for the flexibility that is gained. Currently it's literally impossible to store complex data types in a `Table`, and being able to do so in any way is worth it, in my eyes. It makes `Table` a general data structure for any application instead of a data structure for primitives only.

My current thinking is that I find a spot in the existing class hierarchy that can support a generic `Column` subclass with a minimal implementation, and then try to fill in the missing methods. Ideally this would actually involve making `AbstractColumn<T>` concrete, although that would require a rename.

I don't really see how making an abstract class concrete would break the inheriting classes, as they still belong to the same hierarchy, just with a few more methods provided by default.

**Comment by lwhite1** [comment]
Apparently someone was able to add a BigDecimal column to their table just by subclassing NumberColumn so it may not be as hard as I thought, but I don't understand why AbstractColumn<T> needs to be concrete for this. It seems like you could subclass from it to make a list column type.  

FWIW, the column types I was most interested in when I considered this feature were geographic coordinates, ranges (as in the Guava Range class), monetary values with currency info, and more broadly, support for modeling graphs, which could be done using a ListColumn to create an adjacency list.

**Comment by multimeric** [comment]
I think the reason I wanted a concrete generic column is because I wanted to kind of load any arbitrary tabular file into a tablesaw table without having to think about the data types of each column and create a subclass for each new type, and therefore end up with this huge codebase of all the possible column types I could want. This is especially true if I don't intend to actually *use* any of these columns in any way except to slice them out using another column as an index.
